### ISSUE TITLE
Result returning UNKNOWN the first few times a new regexp is tested

### ISSUE THREAD

**Issue by josdejong** [davisjam/vuln-regex-detector#39]
I noticed that when using both `vulnRegexDetector.test` and `vulnRegexDetector.testSync`, testing a new regexp always returns UNKNOWN the first few times, and only after a minute or so it returns VULNERABLE or SAFE. It would be nice if the library simply waits until the server knows the result and only then return it, right now I cannot reliably use the library in automated unit tests.

**Comment by davisjam** [comment]
This behavior is by design. Testing a regex can take worst-case `O(minutes)/regex * #regexes in line` so I don't see a way to get a response truly synchronously without holding a ton of connections open or paying AWS to scale automatically.

Aside: This is a great fit for serverless if you know anyone giving that away for free.

If you really want to wait for a real response, you can:

1. Run code like this:

```
while (response === UNKNOWN) {
  response = test(...);
}
```

but this will make my server do a bunch of extra work.

2. Install the "application" tools from this repo (`src/detect`, `src/validate`, access via `bin/check-regex.pl`) and run them locally. These tools are written in OCaml (Rathnayake) and Java (Wustholz, Weideman). The `./configure` script will install and build everything so you can run your own queries.

I'm open to other suggestions but I don't see any good alternatives right now.

**Comment by josdejong** [comment]
I understand that testing can require a lot of CPU and that this can get costly. 

I'm looking from a usability point of view though. People using the service do that to get an *answer* to the question whether their regexp is safe or not. If the service returns "UNKNOWN", this is not helpful at all and strongly hurts usability. So people will either have to (a) write a loop to regularly check again until the service returns VULNERABLE/SAFE, or (b) find an other service or application that can answer this question.

If you *do* want this service to be usable in practice I think it's better to solve this issue in the library itself (where you are in control), rather than leave it up to the users to write their own while-loops with unknown intervals (where you don't have control on the impact on your server). I guess you will have to do some testing on what's least expensive: keeping a request open until the test is finished, or regularly poll via a new request until the test is finished.

Ideal would indeed be if people can easily run the tests locally on their own computer. That would mean porting the logic from Java to node.js, or maybe embedding the Java application inside `vuln-regex-detector` and run it via [`node-java`](https://github.com/joeferner/node-java) or something like that.

**Comment by josdejong** [comment]
As for hosting costs: you could simply approach Amazon or Heroku or others, explain your project (making the software world a bit more safe), and ask if they would be interested in giving you hosting for free for this service. Some companies are really positive about these sort of initiatives and eager to lend a helping hand.

**Comment by josdejong** [comment]
Yes, sorry, this doesn't render the service completely unusable.

At my work, alarm bells go off when a CI build fails, you want to prevent that from happening needlessly. For unit tests, integration tests, and CI one aspect I find very important is that it runs reliability and consistently. But if the CI system "sometimes" fails, or fails on a next build for issues of some time ago rather than the changes at hand, that is really bad for the confidence in the CI system. I personally would implement a polling mechanism myself then.

> I've made this as easy as I know how. Run ./configure and then look in bin/ for scripts to run queries against individual regexes, files, directory trees, and git URLs. These are all documented in the project READMEs.

Ah, I hadn't seen that yet in the docs (was looking at the docs on npm which differ from those on github). That sounds promising, thanks. Will give that a try.

**Comment by davisjam** [comment]
See [here for whole repo docs](https://github.com/davisjam/vuln-regex-detector) and [here for bin/](https://github.com/davisjam/vuln-regex-detector/tree/master/bin).

**Comment by josdejong** [comment]
Thanks!

**Comment by davisjam** [comment]
Do let me know your thoughts when you've had a chance to look it over.

**Comment by josdejong** [comment]
ok will do!

**Comment by davisjam** [comment]
It looks like AWS Lambda might be a good fit for this, and the expected workload should fit in the free-or-very-cheap tier. Initial funding could come from the [AWS Research Credits](https://aws.amazon.com/research-credits/) program.

If I understand the architecture right, each lambda could handle a single request, check a backend DB, and either run the detectors locally or return the DB's stored result.
